### ISSUE TITLE
Reference implementation for protocol?

### ISSUE THREAD

**Issue by Ziriax** [Microsoft/VFSForGit#392]
We would like to try VFSForGIT for 3D production. Although we don't have a lot of files, the files are usually large and many of these are binary or un-mergable.

As far as I know, usually Perforce, Subversion or PlasticSCM are used for this. However, since people don't want to clone all files in the repository locally, they manually need to "check-out" just the files or directories they need. Since files can reference other files, this manual operation is error prone, and the GIT atomic mono-repository model seems a lot better, but is not feasible because it would clone unneeded large (gigabytes) files. 

VFSForGIT seems to be great push into the right direction; it would still require some file-locking, but that is already discussed in other threads. 

However, because of the sheer size of 3D movie repositories (terabytes for large productions), using GitHUB or VSTS is not practical, and often a local NAS is used.

However, such a NAS should implement  the VFSForGIT protocol, and at first sight no reference implementation is provided.

Is this something you plan to open source? The protocol is already documented, but say an ASP.NET Core 2 implementation would come very handy :wink:

Or will VFSForGIT not be suitable for this kind of productions?

**Comment by sanoursa** [comment]
No there is no reference implementation of our protocol, but keep in mind that your server would need to implement both that and the Git protocol too. The VFS for Git protocol is just a side channel for talking to a Git server. And given that, most of the difficulty would be in building a Git server that can scale to your needs. As I'm sure you've seen in the protocol doc, the VFS for Git protocol is very simple.

As for your scenario, it's an interesting one but I have no great answers. The core value add of VFS for Git is to solve perf issues for repos with _many_ files. It does also help with large files, but I suspect you'll run into much bigger problems server side in managing a Git repo that large.

**Comment by Ziriax** [comment]
Thanks for reply.  

So this doesn't seem very obvious to do then.  

Typically these 3D repositories do not contain many files (at least compared to software projects), but few large files, and only a few of these are needed for each different user, so the lazy download really helps. For example, it doesn't make sense for a 3D texture painter to download the video files used in editing, or the scene files of the animator, etc.

So I guess after all these years, ClearCase is still the only option (although I don't know if it supports on-demand file-fetching to be honest)

<sup>PS: 16 years ago I actually developed a custom source control system (Evolver) for videogame development, using all the features NTFS gave me for free: permissions for file locking, object-IDs, filesystem tunneling, the change journal, junctions and hardlinks to allow the same file and folder in multiple locations, alternative datastreams for sticky metadata. Every file revision was stored on a file-server using its hash as a filename, and a set of revisions formed a single atomic repository version. Without NTFS this would have not been feasible, but I never had the time to develop a mini-filter to implement lazy download as is done in this amazing project. Unfortunately it was closed source, and the company doesn't exist anymore. Nevertheless, NTFS really rocked for these kind of applications 😃 </sup>
